### DISCUSSION TITLE
Pycharm 配置wsl conda虚拟环境解释器后第三方库代码缺少智能提示

### DISCUSSION THREAD

**Discussion by Physikmann** [physics-data/faq#58] [Q&A] [answered]
系统环境：Windows 10 19042.1165， wsl2 运行 Debian（Debian并非MS中安装，而是官网下载）
软件环境：Windows 10 中使用Pycharm 2021.2 Professional， Debian中安装了Anaconda3 64bit。Debian中存在python2.7的系统底层python，python3.8的conda base python和python3.8的conda虚拟环境python。conda base python3.8 利用conda工具在base环境下下载了numpy和matplotlib，其它均未做任何改动。ssh和git完全正常。

问题描述：当我尝试在Window10中打开wsl debian中的项目，在项目设置中将python解释器定位到debian中的conda虚拟环境解释器（python 3.8），并将工作区定位到每次作业项目所在的文件夹后，编写含有numpy相关方法的代码，正确配置configuration并添加参数后，程序可以正常运行并且输出理想结果，但是在编写程序的过程中，numpy以及其它通过最新conda/pip/pip3安装的第三方库不能够正确弹出智能提示，并且不能够利用 Ctrl + click定位到相关代码。在File中选择Invalid Cache、重启、重装Pycharm、降级Pycahrm、重装wsl、重装Debian、不用虚拟环境、将site-packages安装到root或者user环境均无法结局以上问题。这个给代码编写带来了相当大的麻烦。

问题补充：
常常出现的一条报错信息：
failed to run generator3/__main__.py for wsl://Debian/usr/local/anaconda3/bin/python3.8, exit code 1, stderr: ----- Failed to process '_ast' while nothing yet: [Errno 1] Operation not permitted: '/mnt/c/Users/Administrator/AppData/Local/JetBrains/PyCharm2021.2/python_stubs/1342843747/atexit.py' Traceback (most recent call last): File "/mnt/c/Program Files/JetBrains/PyCharm 2021.2/plugins/python/helpers/generator3/__main__.py", line 167, in   main() File "/mnt/c/Program Files/JetBrains/PyCharm 2021.2/plugins/python/helpers/generator3/__main__.py", line 140, in main generator.discover_and_process_all_modules(name_pattern=args.name_pattern, builtins_only=args.builtins_only) File "/mnt/c/Program Files/JetBrains/PyCharm 2021.2/plugins/python/helpers/generator3/core.py", line 439, in discover_and_process_all_modules self.process_module(mod.qname, mod.path) File "/mnt/c/Program Files/JetBrains/PyCharm 2021.2/plugins/python/helpers/generator3/core.py", line 513, in process_module status = self.reuse_or_generate_skeleton(mod_name, mod_path, sdk_skeleton_state) File "/mnt/c/Program Files/JetBrains/PyCharm 2021.2/plugins/python/helpers/generator3/core.py", line 571, in reuse_or_generate_skeleton copy_skeletons(mod_cache_dir, self.output_dir, get_module_origin(mod_path, mod_name)) File "/mnt/c/Program Files/JetBrains/PyCharm 2021.2/plugins/python/helpers/generator3/util_methods.py", line 780, in copy_skeletons copy(src_dir, dst_dir, merge=True, File "/mnt/c/Program Files/JetBrains/PyCharm 2021.2/plugins/python/helpers/generator3/util_methods.py", line 725, in copy copy(child_src, child_dst, merge=merge, File "/mnt/c/Program Files/JetBrains/PyCharm 2021.2/plugins/python/helpers/generator3/util_methods.py", line 736, in copy shutil.copy(src, dst) File "/usr/local/anaconda3/lib/python3.8/shutil.py", line 416, in copy copymode(src, dst, follow_symlinks=follow_symlinks) File "/usr/local/anaconda3/lib/python3.8/shutil.py", line 305, in copymode chmod_func(dst, stat.S_IMODE(st.st_mode)) PermissionError: [Errno 1] Operation not permitted: '/mnt/c/Users/Administrator/AppData/Local/JetBrains/PyCharm2021.2/python_stubs/1342843747/atexit.py' -----

----------------------------------------------------------------------------------------------------
已经搜索遍Google，StackOverflow，CSDN，各类云论坛均未找到结果，希望有希望解决...

**Comment by Harry-Chen** [comment]
1. 请用代码块正确粘贴错误信息，即用两行 \`\`\` 包围
2. 请尝试不要使用 Anaconda

**Reply by Physikmann** [reply]
不使用anaconda也会产生相同的错误

**Comment by jiegec** [comment]
可能是 PyCharm 的问题，可以去 Jetbrains 的 bug tracker 提交一下

**Comment by heroxbd** [comment, marked as answer]
教学团队的经验和理念不推荐使用 PyCharm 和 Conda。原因如下：

1. PyCharm 是 JetBriains 发售的 IDE （集成开发环境）类工具，有丰富的辅助功能。这些功能是两刃剑。开发者获得便利的同时，会对 IDE 环境进而对 JetBrains 公司产生依赖，不利于发展今后对不同的环境适应能力。课程的教学设计是使用程序文本编辑器，配合终端命令行环境完成程序的创作与调试，不建议使用 IDE。
2. Anaconda 是 Continuum Analytics （现更名为 AnaConda Inc.）发行的商业 Python 环境，凭借资本运作在工业培训领域影响较广。该环境的特点是：有丰富的但是质量低下的软件包，有形似但是缺少设计和基础功能同时性能糟糕的“软件包管理器”，以及看似社区自由分发实则商业封闭运作的软件仓库。课程中的所有软件依赖都不需要通过 Anaconda 取得，对其产生依赖后不利于“最佳工具”品味的培养。

教学团队不建议使用 IDE 和伪发行工具，对它了解和支持也有限，只能给出模糊的建议。如果你必须使用以上工具，需要花精力解决遇到的问题。

**Reply by Physikmann** [reply]
好的，￼十分感谢！
